### DISCUSSION TITLE
Applicative Record Syntax

### DISCUSSION THREAD

**Discussion by Ericson2314** [ghc-proposals/ghc-proposals#472] [Ideas]
I hate being the one asking for new ad-hoc syntax, but, I can't think of another solution.

I am thinking something like:
```haskell
Foo
  { a <- foo
  , b <- bar
  , c <- baz
  }
```
which desugars too:
```haskell
(\a b c -> Foo { a b c }) <$> foo <*> bar <*> baz
```
The goals are are:

1. Error if I forget a field
2. Don't need to worry about order
   - field names should be syntactically local, next to arguments
   - field names should not need to be repeated in a way that could go wrong

I would be open to TH for this. (Hell, I would be open for replacing `do`-notation with TH.) But the TH would be nicer if we had something like `syntax-case` (See Rusts' pattern macros which are merely like `syntax-rules`, but show how to generalize the pattern part to non-sexp) but we don't.

**Comment by JakobBruenker** [comment]
I've always liked [Idris's `!`](https://idris2.readthedocs.io/en/latest/tutorial/interfaces.html#notation), which solves a whole host of similar problems as well as this one. Your example would become
```haskell
Foo {a = !foo, b = !bar, c = !baz}
```
It seems like a bit of a waste to introduce syntax specifically for this when something like `!` would be a much more general solution.

> I would be open to TH for this.

Incidentally, I'm doing some preliminary work to be able to use `!` inside a QuasiQuoter, though of course having things built into the language always comes with some additional amenities.

**Reply by Ericson2314** [reply]
Yeah idiom brackets are nicer

**Reply by JakobBruenker** [reply]
You might know this, but note that it's more general than idiom brackets (Idris has both), since it can be also be nested as in things like `print (read !getLine + !fetchFromDB (MkProduct (read !getLine)))`. So it can also be used for monadic computations rather than just applicative ones.

**Reply by Ericson2314** [reply]
Oh I see. I gathered one didn't need the "bracket" part, but I didn't know one could do that.

Now am I curious, it is better behaved that `ApplicativeDo`, which is very tricky to ensure one doesn't get pessimistic too-monadic stuff?

**Reply by Ericson2314** [reply]
If so, damn I kinda feel like we are a bunch of masochists for not adding that years ago to Haskell.

**Reply by JakobBruenker** [reply]
Hmm it looks like they avoid that rabbit hole by always requiring `Monad` even where `Applicative` would be sufficient, which is I suppose part of the reason they also have Idiom brackets :-/

**Comment by adamgundry** [comment]
I tend to solve this with `RecordWildCards`, `-Werror=missing-fields` (and `ApplicativeDo`, if necessary):
```hs
data Foo = Foo { fld1 :: T, fld2 :: T, fld3 :: T }

do
  fld1 <- foo
  fld2 <- bar
  fld3 <- baz
  pure Foo{..}
```

**Reply by Ericson2314** [reply]
Oh! While I generally try to stay away from both of those, but that's *very* compelling.... :thinking: :thinking: :thinking:
